cerebrium/hardware/using-cuda.mdx

```diff
@@ -0,0 +1,46 @@
+---
+title: Using CUDA
+---
+
+## Overview
+
+CUDA (Compute Unified Device Architecture) enables apps to use graphics cards (GPUs) to speed up calculations. Unlike standard processors (CPUs) that handle one task at a time, graphics cards can handle many tasks simultaneously.
+
+## How CUDA Works
+
+CUDA connects apps to graphics cards, splitting large tasks into smaller pieces that can be processed at the same time. This makes operations like image processing and complex mathematical calculations much faster than using a standard processor alone.
+
+## Using CUDA with app dependencies
+
+Many python packages come with built-in CUDA support. For example, the popular machine learning package _PyTorch_ includes CUDA in its installation:
+
+```toml
+[cerebrium.dependencies.pip]
+torch = "latest" # PyTorch with graphics card support
+```
+
+## Special Requirements
+
+Some apps need direct access to CUDA system libraries and tools. The CUDA base image provides this complete CUDA toolkit environment. This is often necessary when apps:
+
+- Compile custom CUDA code
+- Access low-level CUDA features
+- Need specific CUDA driver versions
+- Require CUDA development tools
+
+The base image can be set in the `cerebrium.toml` file:
+
+```toml
+[cerebrium.deployment]
+docker_base_image_url = "nvidia/cuda:12.1.1-runtime-ubuntu22.04"
+```
+
+## Docker image selection
+
+When building GPU apps, choosing the right Docker base image is crucial for balancing functionality and performance. While the default CUDA runtime image (`nvidia/cuda:12.1.1-runtime-ubuntu22.04`) provides essential CUDA capabilities, some apps may require the expanded toolset found in the `nvidia/cuda:12.1.1-runtime-ubuntu22.04` image or other specialized configurations for the development of new apps. Through the `docker_base_image_url` setting in cerebrium.toml, it is possible to specify any public Docker image that matches the app's requirements, giving flexibility to customize the build environment.
```

Comment:
that isn't our default image. default is 		BaseImage:                  "debian:bookworm-slim",
for all images/hardware etc

Reply:
and we don't allow ANY public image. Only 

	// Allowed base images:
	// - Any 'nvidia/cuda:' image that is also Ubuntu
	// - Any 'debian:' image
	// - Any 'python:' image based on Debian bullseye or bookworm
	// - Specific images
	allowedSpecificImages := map[string]bool{
		"debian:bookworm-slim": true,
		"public.ecr.aws/neuron/pytorch-inference-neuronx:2.1.2-neuronx-py310-sdk2.20.0-ubuntu20.04": true,
		"registry.cerebrium.ai/daily:latest":  true,
		"registry.cerebrium.ai/rimeai:latest": true,
	}

Reply:
A further note, you do not have to just use `nvidia/cuda` to get images to work. These are "kitchen sink" images which include everything. You could for instance, if you wanted to minimise your image size, start with plain `debian` or `ubuntu`, and then add what you need CUDA wise (this is ofc much harder, but doable)